### PR TITLE
chore(infra): mac-os builds fail, solve by porting some changes from main

### PR DESCRIPTION
**Pull request by talalashraf**

### CODE REVIEW

**Review by sdaveas — APPROVED** [review]

File: .github/workflows/build-docker-image-and-binaries.yaml [hunk 1]
```diff
@@ -63,6 +63,39 @@ jobs:
               TOFND_PATH="/Users/runner/work/tofnd/tofnd/target/release"
           fi
 
+      - name: build darwin binaries
```
Comment (milapsheth) [inline .github/workflows/build-docker-image-and-binaries.yaml:66]:
```suggestion
      - name: Build mac os binaries
```

File: .github/workflows/build-docker-image-and-binaries.yaml [hunk 2]
```diff
@@ -63,6 +63,39 @@ jobs:
...
+            cargo build --release --target aarch64-apple-darwin
+            mkdir -p tofndbin
+            mv /Users/runner/work/tofnd/tofnd/target/aarch64-apple-darwin/release/tofnd "./tofndbin/tofnd-$OS-$ARCH-$SEMVER"
+          else
+            export RUSTFLAGS="-L/opt/homebrew/lib"
+            export LIBRARY_PATH=/opt/homebrew/lib:$LIBRARY_PATH
+            export LD_LIBRARY_PATH=/opt/homebrew/lib:$LD_LIBRARY_PATH
+            export C_INCLUDE_PATH=/opt/homebrew/include:$C_INCLUDE_PATH
+            export CPLUS_INCLUDE_PATH=/opt/homebrew/include:$CPLUS_INCLUDE_PATH
+            export HOMEBREW_NO_INSTALL_CLEANUP=TRUE
```
Comment (milapsheth) [inline .github/workflows/build-docker-image-and-binaries.yaml:91]: Duplication between the two if statements. Run the common cmds at the top

File: .github/workflows/build-docker-image-and-binaries.yaml [hunk 3]
```diff
@@ -63,6 +63,39 @@ jobs:
...
+            brew uninstall --ignore-dependencies gmp
+            ARM_DEPENDENCY=$(brew fetch --force --bottle-tag=arm64_sonoma gmp | grep Downloaded | awk '{print $3}')
+            brew install "$ARM_DEPENDENCY"
+            cargo install --locked --path .
+            mkdir -p tofndbin
+            mv "/Users/runner/work/tofnd/tofnd/target/release/tofnd" "./tofndbin/tofnd-$OS-$ARCH-$SEMVER"
```
Comment (milapsheth) [inline .github/workflows/build-docker-image-and-binaries.yaml:97]: Run common cmds at the end

File: .github/workflows/build-docker-image-and-binaries.yaml (hunk 1 above)
Comment (talalashraf) [reply .github/workflows/build-docker-image-and-binaries.yaml:66]: let's keep consistent across repos.

File: .github/workflows/build-docker-image-and-binaries.yaml (hunk 3 above)
Comment (talalashraf) [reply .github/workflows/build-docker-image-and-binaries.yaml:92]: different paths.

File: .github/workflows/build-docker-image-and-binaries.yaml (hunk 2 above)
Comment (talalashraf) [reply .github/workflows/build-docker-image-and-binaries.yaml]: done

File: .github/workflows/build-docker-image-and-binaries.yaml [hunk 4]
```diff
@@ -63,6 +63,34 @@ jobs:
...
+        if: matrix.os == 'macos-latest'
+        run: |
+          OS="darwin"
+          ARCH="${{ matrix.arch }}"
+          if [ "$ARCH" == "arm64" ]
+          export LIBRARY_PATH=/opt/homebrew/lib:$LIBRARY_PATH
+          export HOMEBREW_NO_INSTALL_CLEANUP=TRUE
+          export LD_LIBRARY_PATH=/opt/homebrew/lib:$LD_LIBRARY_PATH
+          brew uninstall --ignore-dependencies gmp
+          then
```
Comment (milapsheth) [inline .github/workflows/build-docker-image-and-binaries.yaml:78]: syntax is broken. shouldn't the if check be here?

File: .github/workflows/build-docker-image-and-binaries.yaml [hunk 5]
```diff
@@ -63,6 +63,34 @@ jobs:
...
+          else
+            export RUSTFLAGS="-L/opt/homebrew/lib"
+            export C_INCLUDE_PATH=/opt/homebrew/include:$C_INCLUDE_PATH
+            export CPLUS_INCLUDE_PATH=/opt/homebrew/include:$CPLUS_INCLUDE_PATH
+            ARM_DEPENDENCY=$(brew fetch --force --bottle-tag=arm64_sonoma gmp | grep Downloaded | awk '{print $3}')
+            brew install "$ARM_DEPENDENCY"
+            cargo install --locked --path .
+            mkdir -p tofndbin
+            mv "/Users/runner/work/tofnd/tofnd/target/release/tofnd" "./tofndbin/tofnd-$OS-$ARCH-$SEMVER"
+          fi
```
Comment (milapsheth) [inline .github/workflows/build-docker-image-and-binaries.yaml:93]:
```suggestion

          export LIBRARY_PATH=/opt/homebrew/lib:$LIBRARY_PATH
          export HOMEBREW_NO_INSTALL_CLEANUP=TRUE
          export LD_LIBRARY_PATH=/opt/homebrew/lib:$LD_LIBRARY_PATH

          brew uninstall --ignore-dependencies gmp
          mkdir -p tofndbin

          if [ "$ARCH" == "arm64" ]; then
            ARM_DEPENDENCY=$(brew fetch --force --bottle-tag=arm64_sonoma gmp | grep Downloaded | awk '{print $3}')

            rustup target add aarch64-apple-darwin
            cargo build --release --target aarch64-apple-darwin

            mv /Users/runner/work/tofnd/tofnd/target/aarch64-apple-darwin/release/tofnd "./tofndbin/tofnd-$OS-$ARCH-$SEMVER"
          else
            export RUSTFLAGS="-L/opt/homebrew/lib"
            export C_INCLUDE_PATH=/opt/homebrew/include:$C_INCLUDE_PATH
            export CPLUS_INCLUDE_PATH=/opt/homebrew/include:$CPLUS_INCLUDE_PATH
            AMD64_DEPENDENCY=$(brew fetch --force --bottle-tag=amd64_sonoma gmp | grep Downloaded | awk '{print $3}')

            brew install "$AMD64_DEPENDENCY"
            cargo install --locked --path .

            mv "/Users/runner/work/tofnd/tofnd/target/release/tofnd" "./tofndbin/tofnd-$OS-$ARCH-$SEMVER"
          fi
```
please add more newlines between different groups of cmds (setting variables, running cmds etc). bash is already not very readable and this makes it even harder.

Also, shouldn't the `else` clause install amd64 dependency and not arm64?

**Review by milapsheth — CHANGES_REQUESTED** [review]

File: .github/workflows/build-docker-image-and-binaries.yaml (hunk 5 above)
Comment (sdaveas) [reply .github/workflows/build-docker-image-and-binaries.yaml]: > Also, shouldn't the else clause install amd64 dependency and not arm64?

I didn't find an `amd64` dep for the gmp lib (looked [here](https://formulae.brew.sh/api/formula/gmp.json)). But it seems that it's able to compile without specifying an architecture manually. Tested it [here](https://github.com/axelarnetwork/tofnd/actions/runs/9374868322), in https://github.com/axelarnetwork/tofnd/pull/329

File: .github/workflows/build-docker-image-and-binaries.yaml [hunk 6]
```diff
@@ -63,6 +63,35 @@ jobs:
               TOFND_PATH="/Users/runner/work/tofnd/tofnd/target/release"
           fi
 
+      - name: build darwin binaries
```
Comment (milapsheth) [inline .github/workflows/build-docker-image-and-binaries.yaml:66]:
```suggestion
      - name: Build Mac OS binaries
```

File: .github/workflows/build-docker-image-and-binaries.yaml [hunk 7]
```diff
@@ -63,6 +63,35 @@ jobs:
...
+
+            rustup target add aarch64-apple-darwin
+
+            cargo build --release --target aarch64-apple-darwin
+            mv /Users/runner/work/tofnd/tofnd/target/aarch64-apple-darwin/release/tofnd "./tofndbin/tofnd-$OS-$ARCH-$SEMVER"
+          else
+            cargo install --locked --path .
+            mv "/Users/runner/work/tofnd/tofnd/target/release/tofnd" "./tofndbin/tofnd-$OS-$ARCH-$SEMVER"
+          fi
       - name: build linux binaries
```
Comment (milapsheth) [inline .github/workflows/build-docker-image-and-binaries.yaml:95]:
```suggestion
      - name: Build Linux binaries
```

**Review by milapsheth — APPROVED** [review]

**Pull request merged**